pkg/docker/config/config.go

```diff
@@ -216,12 +217,29 @@ func getAuthFilePaths(sys *types.SystemContext, homeDir string) []authPath {
 // file or .docker/config.json, including support for OAuth2 and IdentityToken.
 // If an entry is not found, an empty struct is returned.
 func GetCredentials(sys *types.SystemContext, registry string) (types.DockerAuthConfig, error) {
-	return getCredentialsWithHomeDir(sys, registry, homedir.Get())
+	return getCredentialsWithHomeDir(sys, nil, registry, homedir.Get())
+}
+
+// GetCredentialsWithRef returns the registry credentials stored in the
+// registry-specific credential helpers or in the default global credentials
+// helpers with falling back to using either auth.json
+// file or .docker/config.json, including support for OAuth2 and IdentityToken.
+// Compared to GetCredentials, this function accepts a full ref, which will
+// be used for supporting multiple paths when looking up the credentials in
+// auth.json. Path support means, that a given ref
+// "registry.io/some/path/image:latest" will lookup:
+// - registry.io/some/path/image
+// - registry.io/some/path
+// - registry.io/some
+// - registry.io (fallback for compatibility with GetCredentials)
+// If an entry is not found, an empty struct is returned.
+func GetCredentialsWithRef(sys *types.SystemContext, ref reference.Named) (types.DockerAuthConfig, error) {
+	return getCredentialsWithHomeDir(sys, ref, reference.Domain(ref), homedir.Get())
 }
 
 // getCredentialsWithHomeDir is an internal implementation detail of GetCredentials,
 // it exists only to allow testing it with an artificial home directory.
-func getCredentialsWithHomeDir(sys *types.SystemContext, registry, homeDir string) (types.DockerAuthConfig, error) {
+func getCredentialsWithHomeDir(sys *types.SystemContext, ref reference.Named, registry, homeDir string) (types.DockerAuthConfig, error) {
```

Comment:
Hum… we now have record priorities. Should we look up per-repo credentials in config files in preference to per-registry credentials in credential helpers, even if the helpers come first in `helpers`?

I’m leaning towards no (we are explicitly not supporting per-repo in helpers now, and if we ever added that, the prioritization would be a mess — and if anyone is writing per-repo credentials to files now, I can’t see much reason to have the per-registry ones not in the same files), bringing this up to confirm with others (and to have a record of our thinking).

Reply:
I'm not sure about this point, maybe @vrothberg can chime in here.

Reply:
SGTM

Reply:
Just to clarify: So we leave it as it is, right?

Reply:
Yes